### ISSUE TITLE
Decorators on static and instance method order

### ISSUE THREAD

**Issue by just-boris** [wycats/javascript-decorators#39]
I din't found in this document description of decoration applying order. But I think that is good to have it standardized.

So, what's the order of decorations, who goes first: class, static or prototype decorators?

Also, babel now applies prototype decorators before static, which causes issues, but that is technically correct, because the spec has nothing about it. See a [Babel REPL](https://babeljs.io/repl/#?experimental=true&evaluate=true&loose=false&spec=false&code=function%20testable%28target%29%20%7B%0A%20%20target.prototype.testable%20%3D%20true%3B%0A%7D%0A%0Afunction%20ifNotTestable%28target%2C%20name%2C%20descriptor%29%20%7B%0A%20%20if%28target.isTestable%28%29%29%20%7B%0A%20%20%20%20descriptor.value%20%3D%20function%28%29%20%7B%7D%3B%0A%20%20%7D%0A%7D%0A%0A%40testable%0Aclass%20Component%20%7B%0A%0A%20%20%20%20isTestable%28%29%20%7B%0A%20%20%20%20%20%20return%20this.testable%3B%0A%20%20%20%20%7D%0A%0A%20%20%20%20%40ifNotTestable%0A%20%20%20%20hadnler%28%29%20%7B%0A%20%20%20%20%20%20console.log%28'I%20should%20not%20be%20called'%29%0A%20%20%20%20%7D%0A%20%20%20%20%0A%7D%0A%0A%28new%20Component%28%29%29.hadnler%28%29) with unexpected result.

**Comment by silkentrance** [comment]
You don't seem to understand how decorators work. There is no issue with the way decorators work. You are calling isTestable on the target prototype, i.e. the prototype of the decorated class.

Please close as invalid/personal learning process.

**Comment by just-boris** [comment]
@silkentrance Well, I fixed the [code example](https://babeljs.io/repl/#?experimental=true&evaluate=true&loose=false&spec=false&code=function%20testable%28target%29%20%7B%0A%20%20target.prototype.testable%20%3D%20true%3B%0A%7D%0A%0Afunction%20ifNotTestable%28target%2C%20name%2C%20descriptor%29%20%7B%0A%20%20var%20originalMehod%20%3D%20descriptor.value%3B%0A%20%20descriptor.value%20%3D%20function%28%29%20%7B%0A%20%20%20%20if%28this.isTestable%28%29%29%20%7B%0A%20%20%20%20%20%20return%20originalMehod.apply%28this%2C%20arguments%29%3B%0A%20%20%20%20%7D%0A%20%20%7D%3B%0A%20%20%0A%7D%0A%0A%40testable%0Aclass%20Component%20%7B%0A%0A%20%20%20%20isTestable%28%29%20%7B%0A%20%20%20%20%20%20return%20this.testable%3B%0A%20%20%20%20%7D%0A%0A%20%20%20%20%40ifNotTestable%0A%20%20%20%20hadnler%28%29%20%7B%0A%20%20%20%20%20%20console.log%28'I%20should%20not%20be%20called'%29%0A%20%20%20%20%7D%0A%20%20%20%20%0A%7D%0A%0A%28new%20Component%28%29%29.hadnler%28%29), now it is called on instance.

Is it really changing something?

**Comment by silkentrance** [comment]
@just-boris yes, this is correct, now you just have to fix your logic. Hadlner should be called when testable, right? So, it actually is testable as it is being called and will throw the exception.

**Comment by just-boris** [comment]
@silkentrance well, thank you that you pointed that my example is not clear.
I have rewritten it from the scratch: [see link](https://babeljs.io/repl/#?experimental=true&evaluate=true&loose=false&spec=false&code=var%20debuggable%20%3D%20Symbol%28'debuggable'%29%3B%0A%0A%2F%2F%20decorator%20that%20turns%20on%20debugging%20on%20some%20class%0Afunction%20debugEnabled%28target%29%20%7B%0A%20%20target.prototype%5Bdebuggable%5D%20%3D%20true%3B%0A%7D%0A%0A%2F%2F%20decorator%20that%20marks%20method%20as%20intended%20to%20debug%0Afunction%20onlyInDebug%28target%2C%20name%2C%20descriptor%29%20%7B%0A%20%20%2F%2Freplace%20method%20to%20noop%20if%20debug%20is%20not%20enabled%0A%20%20if%28!target%5Bdebuggable%5D%29%20%7B%0A%20%20%20%20descriptor.value%20%3D%20function%28%29%20%7B%7D%0A%20%20%7D%3B%0A%7D%0A%0Aclass%20TestClass%20%7B%0A%20%20%40onlyInDebug%0A%20%20warning%28%29%20%7B%0A%20%20%20%20console.warn%28'This%20warning%20only%20for%20testing'%29%0A%20%20%7D%0A%7D%0Anew%20TestClass%28%29.warning%28%29%3B%20%2F%2F%20no%20warning%2C%20because%20debug%20is%20turned%20off%0A%0A%0A%40debugEnabled%0Aclass%20SecondClass%20%7B%0A%20%20%40onlyInDebug%0A%20%20warning%28%29%20%7B%0A%20%20%20%20console.warn%28'This%20warning%20only%20for%20testing'%29%0A%20%20%7D%0A%7D%0A%0Anew%20SecondClass%28%29.warning%28%29%3B%20%2F%2F%20expected%20to%20get%20a%20warning%2C%20but%20it%20doesn't)

I am going to use the conjunction of two decorators to set up debugging for my classes. First, mark some methods as `onlyInDebug` do disable them in production. Then if you want to debug some classes, just add the `debugEnabled` on top of these classes. As for me, it looks better and flexible alternative rather than the current way to do it with `NODE_ENV` variable.

**Comment by silkentrance** [comment]
@just-boris again, your logic is off. class decorators get applied after that the class was defined.

see

```
var SecondClass = (function () {
  function SecondClass() {
    _classCallCheck(this, _SecondClass);
  }

  _createDecoratedClass(SecondClass, [{
    key: 'warning',
    decorators: [onlyInDebug],
    value: function warning() {
      console.warn('This warning only for testing');
    }
  }]);

  var _SecondClass = SecondClass;
  SecondClass = debugEnabled(SecondClass) || SecondClass;
  return SecondClass;
})();
```

where debugEnabled gets called right before the defined class will be returned.

And before you go and cry out about this being wrong. See for example the way Python decorators work. In fact, the Python interpreter will first call any method / property decorators and will at the very last call upon the class decorators. And this is good, since it has enable the Python community to come up with great frameworks based on that approach, see for example Zope.Interface and what not.

So you need to invert your logic here, marking these methods as in debug only, for example with the symbol and then, when it comes to your class decorator, you will have to eliminate those which should be available during debug only.

But then be sure to also decorate your first class, otherwise it will not work for that, too.

And, perhaps, you should consider trying a different approach?

```
class Foo {
   method() {
        this.debugMethod();
   }

    debugMethod() {
        if (DEBUG) {
          ...
        }
    }
}
```

The above is a well known and commonly accepted pattern.

Alternatively you could rewrite your debugOnly decorator like this

```
function debugOnly(target, attr, descriptor) {
     const result = {...descriptor};
     const ofunc = descriptor.value;
     result.value = function () {
         if (DEBUG) {
             return ofunc.apply(this, arguments);
         }
     };
     return result;
}
```

or something like that.
